airflow/contrib/hooks/bigquery_hook.py

```diff
@@ -1718,6 +1718,97 @@ def get_datasets_list(self, project_id=None):
 
         return datasets_list
 
+    def patch_dataset(self, dataset_id, dataset_resource, project_id=None):
```

Comment:
What do you think about adding the `@GoogleCloudBaseHook.fallback_to_default_project_id``decorator?` This will allow you to delete the repeating code, but also add validation - this field is required.

Reply:
@mik-laj Even though BigQueryHook inherits GoogleCloudBaseHook, most of the functions are in BigQueryBaseCursor. 
I believe that's the reason why when we introduced fallback_to_default_project_id, we didn't put that into bigquery_hook.py.

Reply:
@potiuk introduced this decorator for three reasons:
* to add support for reading the project id from the connection configuration
* to add validation - project id is required parameter
* to remove code duplication.

This decorator was introduced in the AAA class, but is mainly used in classes that inherit from it.
https://github.com/apache/airflow/blob/8dcb9763771957017a2e6fe9db6f67a068463774/airflow/contrib/hooks/gcp_spanner_hook.py
https://github.com/apache/airflow/blob/7cdf99a437534d24b35cf81151bce49a22ef5512/airflow/contrib/hooks/gcp_tasks_hook.py
https://github.com/apache/airflow/blob/8dcb9763771957017a2e6fe9db6f67a068463774/airflow/contrib/hooks/gcp_vision_hook.py
https://github.com/apache/airflow/blob/1805436f3d229f5263dc1f4443da957d7e9b5487/airflow/contrib/hooks/gcp_compute_hook.py
https://github.com/apache/airflow/blob/8dcb9763771957017a2e6fe9db6f67a068463774/airflow/contrib/hooks/gcp_bigtable_hook.py
https://github.com/apache/airflow/blob/1805436f3d229f5263dc1f4443da957d7e9b5487/airflow/contrib/hooks/gcp_sql_hook.py
https://github.com/apache/airflow/blob/8dcb9763771957017a2e6fe9db6f67a068463774/airflow/contrib/hooks/gcp_dlp_hook.py
https://github.com/apache/airflow/blob/1805436f3d229f5263dc1f4443da957d7e9b5487/airflow/contrib/hooks/gcp_function_hook.py
https://github.com/apache/airflow/blob/1805436f3d229f5263dc1f4443da957d7e9b5487/airflow/contrib/hooks/gcp_cloud_build_hook.py
https://github.com/apache/airflow/blob/1805436f3d229f5263dc1f4443da957d7e9b5487/airflow/contrib/hooks/gcp_transfer_hook.py
https://github.com/apache/airflow/blob/1211675651cf9831a2ab8aabbc99eb34a310a643/airflow/contrib/hooks/gcs_hook.py

Recently, my team has added this decorator to official recommendations.
https://docs.google.com/document/d/1_rTdJSLCt0eyrAylmmgYc3yZr-_h51fVlnvMmWqhCkY/edit?ts=5bb72dfd#
The guide will be further developed. I recently was on vacation and did not fully participate in its development. I will soon be adding my comments and changes to complete it

Reply:
@mik-laj I am still confused about your reply. I understood why we introduced the decorator, but it's not quite relevant to my previous comment. 

The decorator can only be used by the functions inside the classes that which inherits `GoogleCloudBaseHook`. So if I put the decorator on my functions, it will not work because my functions are in `BigQueryBaseCursor`.

Reply:
You're right. I withdraw from my idea, but...this decorator will still work here. fallback_to_default_project_id decorator is also compatible with BigQueryBaseCursor class. typehint does not fully describe the requirements. I will try to fix typehint and add decorator here in a separate PR in the future.